src/lib.rs

```diff
@@ -0,0 +1,924 @@
+
+        // We resolve to a scalar value
+        if axes.is_empty() | (remaining_dims.len() == 0) {
+            let max: T = self.data.iter().fold(T::zero(), |acc, x| if acc > *x { acc } else { *x });
+            return Tensor::new(&Shape::new(vec![1]), &[max]);
+        }
+
+        // Create new tensor with right shape
+        let new_shape = Shape::new(remaining_dims);
+        let min: T = self.data.iter().fold(T::zero(), |acc, x| if acc < *x { acc } else { *x });
+        let mut t: Tensor<T> = Tensor::fill(&new_shape, min);
+
+        for target in IndexIterator::new(&new_shape.dims) {
+            let max_iter = IndexIterator::new(&removing_dims);
+            for max_index in max_iter {
+                let mut indices = target.clone();
+                for (i, &axis) in axes.iter().enumerate() {
+                    indices.insert(axis, max_index[i]);
+                }
+
+                if self.get_element(&indices) > t.get_element(&target) {
+                    t.set_element(&target, *self.get_element(&indices));
+                }
+            }
+        }
+
+        t
+    }
+
+    pub fn min(&self, axes: Axes) -> Tensor<T> {
+        let all_axes = (0..self.shape.len()).collect::<Vec<_>>();
+        let remaining_axes = all_axes.clone().into_iter().filter(|&i| !axes.contains(&i)).collect::<Vec<_>>();
+        let remaining_dims = remaining_axes.iter().map(|&i| self.shape.dims[i]).collect::<Vec<_>>();
+        let removing_dims = axes.iter().map(|&i| self.shape.dims[i]).collect::<Vec<_>>();
+
+        // We resolve to a scalar value
+        if axes.is_empty() | (remaining_dims.len() == 0) {
+            let min: T = self.data.iter().fold(T::zero(), |acc, x| if acc < *x { acc } else { *x });
+            return Tensor::new(&Shape::new(vec![1]), &[min]);
+        }
+
+        // Create new tensor with right shape
+        let new_shape = Shape::new(remaining_dims);
+        let max: T = self.data.iter().fold(T::zero(), |acc, x| if acc > *x { acc } else { *x });
+        let mut t: Tensor<T> = Tensor::fill(&new_shape, max);
+
+        for target in IndexIterator::new(&new_shape.dims) {
+            let min_iter = IndexIterator::new(&removing_dims);
+            for min_index in min_iter {
+                let mut indices = target.clone();
+                for (i, &axis) in axes.iter().enumerate() {
+                    indices.insert(axis, min_index[i]);
+                }
+
+                if self.get_element(&indices) < t.get_element(&target) {
+                    t.set_element(&target, *self.get_element(&indices));
+                }
+            }
+        }
+
+        t
+    }
+
+    /// For the maths see: https://bit.ly/3KQjPa3
+    fn calculate_index(&self, indices: &[usize]) -> usize {
+        let mut index = 0;
+        for k in 0..self.shape.len() {
+            let stride = self.shape.dims[k+1..].iter().product::<usize>();
+            index += indices[k] * stride;
+        }
+        index
+    }
```

Comment:
This would potentially be very useful to abstract away for other Tensor types

Reply:
You mean like sparse tensors?

Reply:
Sparse and static.

Reply:
Actually mostly for static as sparse tensors are a different beast

Reply:
Actually, this could work well - why don't we make this a method for the Shape struct. Since technically all you need is a shape to work out the right index?

Reply:
We agreed that this should go in a `RawTensor` struct which is responsible for laying out the tensor in memory.